tests/integration/test_traefik_deployed_after_metallb.py

```diff
@@ -82,14 +100,87 @@ async def test_build_and_deploy(ops_test: OpsTest, traefik_charm):
+        logger.debug("Attempting to reach %s", ep)  # Traceback doesn't spell out the endpoint
+        urlopen(ep)
+        # A 404 would result in an exception:
+        #   urllib.error.HTTPError: HTTP Error 404: Not Found
+        # so just `urlopen` on its own should suffice for the test.
+
+
+async def curl_endpoints(ops_test: OpsTest, certs_dir, cert_path, traefik_app_ip):
+    for endpoint in get_endpoints(ops_test, scheme="https", netloc=mock_hostname):
+        # Tell curl to resolve the mock_hostname as traefik's IP (to avoid using a custom DNS
+        # server). This is needed because the certificate issued by the CA would have that same
+        # hostname as the subject, and for TLS to succeed, the target url's hostname must match
+        # the one in the certificate.
+        rc, stdout, stderr = await ops_test.run(
+            "curl",
+            "-s",
+            "--fail-with-body",
+            "--resolve",
+            f"{mock_hostname}:443:{traefik_app_ip}",
+            "--capath",
+            certs_dir,
+            "--cacert",
+            cert_path,
+            endpoint,
+        )
+        logger.info("%s: %s", endpoint, (rc, stdout, stderr))
+        assert rc == 0, (
+            f"curl exited with rc={rc} for {endpoint}; "
+            "non-zero return code means curl encountered a >= 400 HTTP code"
+        )
+
+
+@pytest.mark.abort_on_fail
+async def test_tls_termination(ops_test: OpsTest, temp_dir):
+    # TODO move this to the bundle tests
+    await ops_test.model.applications[trfk.name].set_config({"external_hostname": mock_hostname})
+
+    await ops_test.model.deploy(
+        "ch:tls-certificates-operator",
+        application_name="root-ca",
+        channel="edge",
+    )
+    await ops_test.model.applications["root-ca"].set_config(
+        {
+            "ca-common-name": "demo.ca.local",
+            "generate-self-signed-certificates": "true",
+        }
+    )
+    await ops_test.model.add_relation("root-ca", f"{trfk.name}:certificates")
+    await ops_test.model.wait_for_idle(status="active", timeout=300)
+
+    # Get self-signed cert from peer app data
+    rc, stdout, stderr = await ops_test.run("juju", "show-unit", "root-ca/0", "--format=json")
+    data = json.loads(stdout)
+    peer_data = next(
+        filter(lambda d: d["endpoint"] == "replicas", data["root-ca/0"]["relation-info"])
+    )
+    cert = peer_data["application-data"]["self_signed_ca_certificate"]
+
+    cert_path = temp_dir / "local.cert"
+    with open(cert_path, "wt") as f:
+        f.writelines(cert)
+
+    ip = await get_address(ops_test, trfk.name)
+    await curl_endpoints(ops_test, temp_dir, cert_path, ip)
+
+
+@pytest.mark.abort_on_fail
+async def test_tls_termination_after_charm_upgrade(ops_test: OpsTest, traefik_charm, temp_dir):
+    logger.info(
+        "Refreshing charm to test TLS termination still works with the same certificate after"
+        " charm upgrade..."
```

Comment:
Not sure this is the desired behavior: does it matter that traefik's cert changed?
As long as the root-ca cert is the same, https should keep working.

Reply:
This is a bad assumption. Unless the user actually imports the root CA into the store (browser or otherwise), most browsers these days do (or did, probably still do -- there were lots of debates about it in the Mozilla community) bring up a warning page again if a self-signed cert changed for any reason.

Reply:
Ok sure we can keep the current behavior then. We can deal with stored state later.

Reply:
I see no issue with having to re-accept a self signed cert. This seems like an acceptable burden if you are using self signed certs

Reply:
It is, generally. There's a case here where, because Traefik will use an auto-generated self-signed cert if it doesn't have any, connecting to Traefik will give you an entirely different cert chain, then relating to a TLS provider (which grants a cert) will override the entire cert chain (signing/root CAs and all).

Since we aren't using a middleware to upgrade the connection, users would kind-of need to opt-in (unless they deployed an application which set HSTS, but then that's on them), so programmatic clients _probably_ wouldn't care. _Most_ clients (urllib, requests, tls.Client in golang, etc don't cache certs).